### PR TITLE
Add support for breakpoints on SystemTap probes

### PR DESCRIPTION
**Pull request by Bob131**
Sorry, this is kind of a big one. Some notes:
 - I split this into two commits: one adding librraudit, and another actually implementing stap probe support. The auditing library may be useful in its own right for future work, so it seemed logical to break that out into a separate commit.
 - The implementation of `AddressSpace::add_stap_semaphore_range` is kind of ugly. I do have a version that implemented the semaphore data structure as the minimal set of disjoint intervals, but the code ended up looking almost comically over-complicated given the objective it fulfilled.
 - I'm not sure how you might want the test suite altered to exercise this. FWIW, I did run the suite with `--stap-sdt` and it passed with the exception of one test (which failed due to a watchpoint on mmap'd memory being triggered when the same address was mmap'd again by the audit library)

### CODE REVIEW

File: src/audit/rtld-audit.c [hunk 1]
```diff
@@ -43,3 +51,166 @@ unsigned la_version(unsigned version) {
...
+    semaphore_addr_range_init_single(range, address);
+  }
+}
+
+unsigned la_objopen(struct link_map* map, Lmid_t lmid, uintptr_t* cookie) {
+  StapNoteIter iter;
+  ElfStapNote note;
+  SemaphoreAddrRange range;
+
+  (void) cookie;
```
Comment (rocallahan) [inline src/audit/rtld-audit.c]: Use `__attribute__((unused))` instead?

File: src/audit/stap-note-iter.c [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    return NULL;
+  }
+
+  return (char*) map + slack;
+}
+
+static void stap_note_iter_unmap(StapNoteIter* self, void* data, size_t size) {
+  uintptr_t data_addr = (uintptr_t) data;
+  void* page = (void*) (data_addr & ~(sysconf(_SC_PAGE_SIZE) - 1));
+  (void) self;
```
Comment (rocallahan) [inline src/audit/stap-note-iter.c]: `__attribute__((unused))`

**Comment by rocallahan** [comment]
Sorry I've only got nits here, I didn't check the audit library code in much detail.

File: src/Monkeypatcher.cc [hunk 3]
```diff
@@ -59,33 +60,38 @@ static void write_and_record_mem(RecordTask* t, remote_ptr<T> child_addr,
...
-template <typename Arch> static void setup_preload_library_path(RecordTask* t) {
-  static_assert(sizeof(SYSCALLBUF_LIB_FILENAME_PADDED) ==
-                    sizeof(SYSCALLBUF_LIB_FILENAME_32),
-                "filename length mismatch");
+#define setup_library_path(arch, env_var, soname, task) \
+  _setup_library_path<arch>(task, env_var, soname ## _BASE, \
+                            soname ## _PADDED, soname ## _32)
 
+template <typename Arch>
+static void _setup_library_path(RecordTask* t, const char* env_var,
```
[the remark this answers is not in the document]
Comment (Bob131) [reply src/Monkeypatcher.cc]: Wouldn't that make it a macro invocation?

File: src/audit/rtld-audit.c (hunk 1 above)
Comment (Bob131) [reply src/audit/rtld-audit.c]: Ack.

File: src/audit/stap-note-iter.c (hunk 2 above)
Comment (Bob131) [reply src/audit/stap-note-iter.c]: Ack.

**Comment by Bob131** [comment]
Pinging, in case you forgot about this over the holiday break ;)

File: src/Monkeypatcher.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (Bob131) [reply src/Monkeypatcher.cc]: Yep, I can do that.

**Comment by rocallahan** [comment]
Sorry, I had pending comments I had failed to submit.

See "Well, just give it another name, e.g. setup_library_path_arch".

**Comment by rocallahan** [comment]
Thanks!

**Comment by Bob131** [comment]
No worries! Thank you for being receptive to such a change

**Pull request merged**